File: 01_Language/01_Functions/python/hash_equals.py
# coding: utf-8

import hmac


def hash_equals(known_string, user_string):
    return hmac.compare_digest(known_string, user_string)


if __name__ == '__main__':
    import crypt

    expected = crypt.crypt("12345", "$2a$07$usesomesillystringforsalt$")
    correct = crypt.crypt("12345", "$2a$07$usesomesillystringforsalt$")
    incorrect = crypt.crypt("apple", "$2a$07$usesomesillystringforsalt$")

    print(hash_equals(expected, correct))
    print(hash_equals(expected, incorrect))

File: 01_Language/01_Functions/java/src/main/java/com/onevgo/functions/HashEquals.java
package com.onevgo.functions;

import java.security.MessageDigest;

public class HashEquals {
    public static boolean hashEquals(String knownString, String userString) {
        return hashEquals(knownString.getBytes(), userString.getBytes());
    }

    public static boolean hashEquals(byte[] knownString, byte[] userString) {
        return MessageDigest.isEqual(knownString, userString);
    }

    public static void main(String[] args) {
        String expected = Crypt.crypt("12345", "$2a$07$usesomesillystringforsalt$");
        String correct = Crypt.crypt("12345", "$2a$07$usesomesillystringforsalt$");
        String incorrect = Crypt.crypt("apple", "$2a$07$usesomesillystringforsalt$");

        System.out.println(hashEquals(expected, correct));
        System.out.println(hashEquals(expected, incorrect));
    }
}
